File: backend/src/main/java/com/coing/domain/bookmark/service/BookmarkService.java
package com.coing.domain.bookmark.service;

import java.util.List;
import java.util.UUID;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.coing.domain.bookmark.controller.dto.BookmarkRequest;
import com.coing.domain.bookmark.controller.dto.BookmarkResponse;
import com.coing.domain.bookmark.entity.Bookmark;
import com.coing.domain.bookmark.repository.BookmarkRepository;
import com.coing.domain.coin.market.entity.Market;
import com.coing.domain.coin.market.repository.MarketRepository;
import com.coing.domain.user.CustomUserPrincipal;
import com.coing.domain.user.entity.User;
import com.coing.domain.user.repository.UserRepository;
import com.coing.global.exception.BusinessException;
import com.coing.util.MessageUtil;
import com.coing.util.PageUtil;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class BookmarkService {

	private final BookmarkRepository bookmarkRepository;
	private final MarketRepository marketRepository;
	private final UserRepository userRepository;
	private final MessageUtil messageUtil;

	@Transactional
	public void addBookmark(BookmarkRequest request, CustomUserPrincipal principal) {
		UUID userId = principal.id();
		String coinCode = request.coinCode();

		// 기존 북마크 존재 여부 확인
		if (bookmarkRepository.existsByUserIdAndMarketCode(userId, coinCode)) {
			throw new BusinessException(messageUtil.resolveMessage("bookmark.already.exists"),
				HttpStatus.BAD_REQUEST);
		}

		User user = userRepository.findById(userId)
			.orElseThrow(() -> new BusinessException(messageUtil.resolveMessage("member.not.found"),
				HttpStatus.NOT_FOUND));

		Market market = marketRepository.findById(coinCode)
			.orElseThrow(() -> new BusinessException(messageUtil.resolveMessage("market.not.found"),
				HttpStatus.NOT_FOUND));

		Bookmark bookmark = Bookmark.builder()
			.user(user)
			.market(market)
			.build();

		bookmarkRepository.save(bookmark);
		//return BookmarkResponse.of(bookmark);
	}

	@Transactional(readOnly = true)
	public Page<BookmarkResponse> getBookmarksByQuote(UUID userId, String quote, Pageable pageable) {
		List<Bookmark> bookmarks = bookmarkRepository.findByUserIdAndQuote(userId, quote);

		List<BookmarkResponse> responses = bookmarks.stream()
			.map(BookmarkResponse::of)
			.toList();

		return PageUtil.paginate(responses, pageable);
	}

	@Transactional
	public void deleteBookmark(UUID userId, String marketCode) {
		Bookmark bookmark = bookmarkRepository.findByMarketCode(marketCode);
		if (bookmark == null) {
			throw new BusinessException(messageUtil.resolveMessage("bookmark.not.found"),
				HttpStatus.NOT_FOUND);
		}

		// 인증된 사용자가 북마크의 소유자인지 확인
		if (!bookmark.getUser().getId().equals(userId)) {
			throw new BusinessException(messageUtil.resolveMessage("bookmark.access.denied"),
				HttpStatus.FORBIDDEN);
		}

		bookmarkRepository.delete(bookmark);
	}
}

File: backend/src/main/java/com/coing/domain/bookmark/service/BookmarkService.kt
package com.coing.domain.bookmark.service

import com.coing.domain.bookmark.controller.dto.BookmarkRequest
import com.coing.domain.bookmark.controller.dto.BookmarkResponse
import com.coing.domain.bookmark.entity.Bookmark
import com.coing.domain.bookmark.repository.BookmarkRepository
import com.coing.domain.coin.market.repository.MarketRepository
import com.coing.domain.user.dto.CustomUserPrincipal
import com.coing.domain.user.repository.UserRepository
import com.coing.global.exception.BusinessException
import com.coing.util.MessageUtil
import com.coing.util.PageUtil
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class BookmarkService(
    private val bookmarkRepository: BookmarkRepository,
    private val marketRepository: MarketRepository,
    private val userRepository: UserRepository,
    private val messageUtil: MessageUtil
) {

    @Transactional
    fun addBookmark(request: BookmarkRequest, principal: CustomUserPrincipal?) {
        if (principal == null) {
            throw BusinessException(messageUtil.resolveMessage("member.not.found"), HttpStatus.NOT_FOUND)
        }

        val userId = principal.id
        val coinCode = request.coinCode

        // 북마크가 이미 존재하는지 확인
        if (bookmarkRepository.existsByUserIdAndMarketCode(userId, coinCode)) {
            throw BusinessException(messageUtil.resolveMessage("bookmark.already.exists"), HttpStatus.BAD_REQUEST)
        }

        val user = userRepository.findById(userId).orElseThrow {
            BusinessException(messageUtil.resolveMessage("member.not.found"), HttpStatus.NOT_FOUND)
        }

        val market = marketRepository.findById(coinCode).orElseThrow {
            BusinessException(messageUtil.resolveMessage("market.not.found"), HttpStatus.NOT_FOUND)
        }

        val bookmark = Bookmark(
            user = user,
            market = market
        )

        bookmarkRepository.save(bookmark)
    }

    @Transactional(readOnly = true)
    fun getBookmarksByQuote(
        principal: CustomUserPrincipal?,
        quote: String,
        pageable: Pageable
    ): Page<BookmarkResponse> {
        if (principal == null) {
            throw BusinessException(messageUtil.resolveMessage("member.not.found"), HttpStatus.NOT_FOUND)
        }

        val userId = principal.id
        val bookmarks = bookmarkRepository.findByUserIdAndQuote(userId, quote)

        val responses = bookmarks.map { BookmarkResponse.of(it) }

        return PageUtil.paginate(responses, pageable)
    }

    @Transactional
    fun deleteBookmark(userId: UUID, marketCode: String) {
        val bookmark = bookmarkRepository.findByMarketCode(marketCode)
            ?: throw BusinessException(messageUtil.resolveMessage("bookmark.not.found"), HttpStatus.NOT_FOUND)

        // 요청한 유저가 소유자인지 검증
        if (bookmark.user.id != userId) {
            throw BusinessException(messageUtil.resolveMessage("bookmark.access.denied"), HttpStatus.FORBIDDEN)
        }

        bookmarkRepository.delete(bookmark)
    }
}
